File: tests/modules/good/python/dictionnary_ext.py
# This is a template file following the expected interface and declarations to
# implement the corresponding Catala module.
#
# You should replace all `raise Impossible` place-holders with your
# implementation and rename it to remove the ".template" suffix.

from catala_runtime import *
from typing import Any, List, Callable, Tuple
from enum import Enum
from . import Stdlib_en as stdlib_en
from . import Date_en as date_en
from . import MonthYear_en as month_year_en
from . import Period_en as period_en
from . import Money_en as money_en
from . import Integer_en as integer_en
from . import Decimal_en as decimal_en
from . import List_en as list_en

Dictionnary = dict

empty = {}

def store(dict:Dictionnary, key:Integer, value:Money):
    ret = dict.copy()
    ret[key.value] = value
    return ret

def find(dict:Dictionnary, key:Integer):
    if key.value in dict:
        return Option(dict[key.value])
    else:
        return Option(None)

File: tests/modules/good/c/dictionnary_ext.c
/* This is a template file following the expected interface and declarations to
 * implement the corresponding Catala module.
 *
 * You should replace all `catala_error(catala_impossible)` place-holders with
 * your implementation and rename it to remove the ".template" suffix. */

#include <stdio.h>
#include <stdlib.h>
#include <catala_runtime.h>

#include <Stdlib_en.h>
#include <Date_en.h>
#include <MonthYear_en.h>
#include <Period_en.h>
#include <Money_en.h>
#include <Integer_en.h>
#include <Decimal_en.h>
#include <List_en.h>

typedef struct dict {
  CATALA_INT key;
  CATALA_MONEY value;
  struct dict * rest;
} Dictionnary_ext_Dictionnary;

const Dictionnary_ext_Dictionnary* DictionnaryExt__empty () {
  return NULL;
}

const Dictionnary_ext_Dictionnary* DictionnaryExt__store
    (Dictionnary_ext_Dictionnary* dict,
     CATALA_INT key,
     CATALA_MONEY value)
{
  struct dict * head = catala_malloc(sizeof(struct dict));
  head->key = key;
  head->value = value;
  head->rest = dict;
  return head;
}

const CATALA_OPTION(CATALA_MONEY) DictionnaryExt__find
    (Dictionnary_ext_Dictionnary* dict, CATALA_INT key)
{
  while (dict) {
    if (mpz_cmp(key, dict->key)) dict = dict->rest;
    else return catala_some(dict->value);
  }
  return &catala_none;
}
